src/hotspot/share/utilities/quickSort.hpp

```diff
@@ -53,48 +51,43 @@ class QuickSort : AllStatic {
     size_t last_index = length - 1;
 
     if (comparator(array[0], array[middle_index]) > 0) {
-      swap(array, 0, middle_index);
+      swap_elements(array, 0, middle_index);
     }
     if (comparator(array[0], array[last_index]) > 0) {
-      swap(array, 0, last_index);
+      swap_elements(array, 0, last_index);
     }
     if (comparator(array[middle_index], array[last_index]) > 0) {
-      swap(array, middle_index, last_index);
+      swap_elements(array, middle_index, last_index);
     }
     // Now the value in the middle of the array is the median
-    // of the fist, last and middle values. Use this as pivot.
+    // of the first, last and middle values. Use this as pivot.
     return middle_index;
   }
 
-  template<bool idempotent, class T, class C>
+  template<class T, class C>
   static size_t partition(T* array, size_t pivot, size_t length, C comparator) {
     size_t left_index = 0;
     size_t right_index = length - 1;
     T pivot_val = array[pivot];
 
     for ( ; true; ++left_index, --right_index) {
       for ( ; comparator(array[left_index], pivot_val) < 0; ++left_index) {
-        assert(left_index < length, "reached end of partition");
+        assert(left_index < (length - 1), "reached end of partition");
```

Comment:
Let me see if I understand this change. It makes assert stronger: we do not accept `left_index == length - 1` anymore. I guess that would mean the pivot is at the last element? Which makes the partition is empty, which cannot happen?

Reply:
The reason I looked at the assert carefully in the first place was that `left_index < length` could
pass and yet we could still do `array[length]`, which could be UB.  So the tightened condition must
be correct, else we have a bug.  Proving we don't have a bug is a little bit harder.

We know that the last element in the sequence can't be less than the pivot,
because of the way find_pivot works.  It arranges for the first, middle, and
last values in the sequence to be ordered A[0] <= A[pivot] <= A[length-1].

The tricky case is the sequence [Vi..., P, P], where P is the pivot and all
Vi < P. The from-left scan proceeds until it reaches the left occurrence of P.
The from-right scan immediately stops on the right occurrence of P. The two
P's are swapped. left_index is incremented and right_index is decremented.
The from-left scan is restarted with left_index == length - 1, stops
immediately because P < P is false, and does not execute the assert. Similarly,
the from-right scan is restarted with right_index == length - 2, and also stops
because P > P is false.  And now the outer loop terminates.

A similar argument applies for the from-right scan, where we already assert
right_index > 0.